File: app/src/components/ensembleScatterplot/tooltip.js
/**
 * Copyright 2017-2020 Lawrence Livermore National Security, LLC and other
 * CallFlow Project Developers. See the top-level LICENSE file for details.
 * SPDX-License-Identifier: MIT
 */
import * as d3 from "d3";
import * as utils from "../utils";

export default {
	template: `<g id="tooltip_scatterplot"></g>`,
	name: "ToolTip",
	data: () => ({
		id: "",
		textCount: 0,
		textxOffset: 20,
		textyOffset: 20,
		textPadding: 18,
		offset: 10,
		fontSize: 12,
	}),

	methods: {
		init(id) {
			this.id = id;
			this.toolTipDiv = d3.select("#" + this.id)
				.append("svg")
				.attr("class", "toolTipSVG");

			this.toolTipG = this.toolTipDiv.append("g");
			this.height = 80;
			this.halfWidth = document.getElementById(this.id).clientWidth / 2;
			this.halfHeight = document.getElementById(this.id).clientHeight / 2;
		},

		addText(text) {
			this.toolTipText = this.toolTipG
				.append("text")
				.style("font-family", "sans-serif")
				.style("font-size", "")
				.attrs({
					"class": "toolTipContent",
					"x": () => {
						if (this.mousePosX + this.halfWidth > document.getElementById(this.id).clientWidth) {
							return (this.mousePosX - this.width) + this.textxOffset + "px";
						}
						return this.mousePosX + this.textxOffset + "px";
					},
					"y": () => {
						if (this.mousePosY + this.halfHeight > document.getElementById(this.id).clientHeight) {
							return ((this.mousePosY) + this.textyOffset + this.textPadding * this.textCount) - this.height + "px";
						}
						return (this.mousePosY) + this.textyOffset + this.textPadding * this.textCount + "px";
					}
				})
				.text(text);
			this.textCount += 1;
		},

		trunc(str, n) {
			str = str.replace(/<unknown procedure>/g, "proc ");
			return (str.length > n) ? str.substr(0, n - 1) + "..." : str;
		},

		truncTimeLabel(str) {
			if (str == "Inclusive") {
				return "Inc.";
			}
			else if (str == "Exclusive") {
				return "Exc.";
			}
		},

		info() {
			this.addText("Callsite: " + utils.truncNames(this.data.callsite, 10));
			this.addText("Mean " + this.truncTimeLabel(this.$store.selectedMetric) + " Time: " + utils.formatRuntimeWithUnits(this.data.value));
			this.addText("Run: " + this.data.run);
			// this.addText('QCD: ' + this.data.QCD.toFixed(3))
		},

		render(data) {
			this.clear();
			this.width = 19 * this.fontSize;
			var svgScale = d3.scaleLinear().domain([2, 11]).range([50, 150]);
			this.mousePos = d3.mouse(d3.select("#" + this.id).node());
			this.mousePosX = this.mousePos[0];
			this.mousePosY = this.mousePos[1];
			this.toolTipG.attr("height", svgScale(10) + "px");
			this.toolTipRect = this.toolTipG
				.append("rect")
				.attrs({
					"class": "toolTipContent",
					"fill": "white",
					"stroke": "black",
					"rx": "10px",
					"fill-opacity": 1,
					"width": this.width,
					"height": this.height,
					"x": () => {
						if (this.mousePosX + this.halfWidth > document.getElementById(this.id).clientWidth) {
							return (this.mousePosX - this.width) + "px";
						}
						return (this.mousePosX) + "px";
					},
					"y": () => {
						if (this.mousePosY + this.halfHeight > document.getElementById(this.id).clientHeight) {
							return (this.mousePosY - this.height) + "px";
						}
						return (this.mousePosY) + "px";
					}
				});
			this.data = data;
			this.info();
		},

		clear() {
			this.textCount = 0;
			d3.selectAll(".toolTipContent").remove();
		},

	}
};

File: callflow/modules/delete/tooltip.py
# Copyright 2017-2020 Lawrence Livermore National Security, LLC and other
# CallFlow Project Developers. See the top-level LICENSE file for details.
#
# SPDX-License-Identifier: MIT
# ------------------------------------------------------------------------------

import pandas as pd

import callflow
from callflow.utils.utils import df_lookup_by_column, df_names_in_module

LOGGER = callflow.get_logger()


# ------------------------------------------------------------------------------
# ------------------------------------------------------------------------------
class ToolTip:

    KEYS_TO_ADD = ['name', 'rank', 'time', 'time (inc)', 'callers', 'callees']

    def __init__(self, state, module):

        self.state = state
        self.graph = state.new_gf.graph
        self.df = state.new_gf.df
        self.entire_df = state.new_entire_gf.df

        self.module = module
        self.result = self.run()

    def run(self):

        ret = []

        callsites = df_names_in_module(self.df, self.module)
        for func in callsites:

            _ret = {}

            name_entire_df = df_lookup_by_column(self.entire_df, "name", func)
            name_df = df_lookup_by_column(self.df, "name", func)

            for _ in ToolTip.KEYS_TO_ADD:
                if _ == 'name':
                    _ret[_] = func
                elif _ in ['callers', 'callees']:
                    _ret[_] = name_df[_].tolist()
                else:
                    _ret[_] = name_entire_df[_].tolist()

        return pd.DataFrame(ret).to_json(orient="columns")

# ------------------------------------------------------------------------------
